### PR TITLE
feat: teach compactor to compact smaller number of files

### PR DESCRIPTION
**Pull request by NGA-TRAN**
There are 2 main changes in this PR:
1. Limit the number of compacting files per compaction job. If the number of overlapped files are larger than that limit or their size is over the size limit, they will be split further into smaller groups of files to avoid compacting too much in one job.
2. Refactor to group the files of a partition into groups of compacting files and upgraded files before the actual compaction work to make sure we do not allocate resource to compact partitions that do not need to do so.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/handler.rs [hunk 1]
```diff
@@ -170,28 +179,50 @@ async fn run_compactor(compactor: Arc<Compactor>, shutdown: CancellationToken) {
...
-                debug!(candidate=?c, "compaction complete");
-            });
-            used_size += c.file_size_bytes;
-            handles.push(handle);
-            if used_size > max_size {
-                debug!(%max_size, %used_size, n_compactions=%handles.len(), "reached maximum concurrent compaction size limit");
-                break;
+                Ok(compact_and_upgrade) => {
+                    if compact_and_upgrade.compactable() {
+                        let handle = tokio::task::spawn(async move {
```
Comment (NGA-TRAN) [inline compactor/src/handler.rs:204]: Only spawn new Tokio for actual compacting work

File: clap_blocks/src/compactor.rs [hunk 2]
```diff
@@ -36,22 +36,31 @@ pub struct CompactorConfig {
     /// The compactor will limit the number of simultaneous compaction jobs based on the
     /// size of the input files to be compacted.  This number should be less than 1/10th
     /// of the available memory to ensure compactions have
-    /// enough space to run. Default is 100,000,000 (100MB).
+    /// enough space to run. Default is 10,000,000 (10MB).
     #[clap(
         long = "--compaction-concurrent-size-bytes",
         env = "INFLUXDB_IOX_COMPACTION_CONCURRENT_SIZE_BYTES",
-        default_value = "100000000"
+        default_value = "10000000"
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply clap_blocks/src/compactor.rs:43]: I really want to reduce this to 1 (or 2) for us at least to have things run smoothly before increasing the concurrency level. The purpose is to narrow down the problem. Nothing finishes right now.

BTW, the update in #4696 shows number of jobs on 4 compactors respectively: 7, 1, 2, 3

File: compactor/src/compact.rs [hunk 3]
```diff
@@ -150,7 +180,7 @@ pub struct Compactor {
 
     /// Counter for the actual number of files compacted
     /// compaction_actual_counter = compaction_counter - some large-enough non-overlapped files
-    compaction_actual_counter: Metric<U64Counter>,
+    // compaction_actual_counter: Metric<U64Counter>,
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor/src/compact.rs:183]: This is intended. I should have removed this line.This PR removes the unnecessary need to report files that are not compacted

File: compactor/src/compact.rs [hunk 4]
```diff
@@ -748,15 +814,72 @@ impl Compactor {
...
+                    let mut size = 0;
+                    for (i, item) in group.iter().enumerate() {
+                        if i >= max_count {
+                            count = max_count;
+                            break;
+                        }
+
+                        size += item.file_size_bytes;
+                        if size >= max_size_bytes {
+                            count = i + 1;
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor/src/compact.rs:848]: . Yeah, I took a while for me to write this function right. The unit test for this is `test_limit_num_files` but I will modify it a bit os split it to many tests to explicitly call this function rather than its caller. I also modified available tests to test this function.

. Increasing the counter (+1) is needed because the `split_off` function uses `len` instead of `index`

[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor/src/compact.rs:848]: Another reason this function is hard to follow because it consider both num files and size at the same time

**Review by pauldix — APPROVED** [review]
Ok, I think this does what it says. We'll likely need to make the compaction logic more sophisticated to pick up smaller files that can be combined on a later pass, like if 10 files compact to something that's still small, you'd want it to be compacted with another small file later if possible.

File: compactor/src/compact.rs [hunk 5]
```diff
@@ -124,6 +124,36 @@ pub enum Error {
 /// A specialized `Error` for Compactor Data errors
 pub type Result<T, E = Error> = std::result::Result<T, E>;
 
+/// Data of parquet files to compact and upgrade
+pub struct CompactAndUpgrade {
+    // sequencer ID of all files in this struct
+    sequencer_id: Option<SequencerId>,
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor/src/compact.rs:130]: I need to return empty CompactAndUpgrade if the partition has no files

File: compactor/src/compact.rs [hunk 6]
```diff
@@ -1924,13 +2069,163 @@ mod tests {
...
+        // newest files and very large
+        let alone = arbitrary_parquet_file_with_size(30, 35, compaction_max_size_bytes + 200); // too large to group
+        let newest_overlapped_group = vec![alone.clone()];
+
+        // small files in  the middle
+        let another = arbitrary_parquet_file_with_size(13, 15, 1000);
+        let partial_overlap = arbitrary_parquet_file_with_size(14, 16, 2000);
+        let middle_overlapped_group = vec![another.clone(), partial_overlap.clone()];
+
+        // Note: the
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply compactor/src/compact.rs:2109]: I will remove this

**Comment by NGA-TRAN** [comment]
@pauldix 
> We'll likely need to make the compaction logic more sophisticated to pick up smaller files that can be combined on a later pass, like if 10 files compact to something that's still small, you'd want it to be compacted with another small file later if possible.

Yes, a good one for future improvements. One thing we can do is size related `quorum`. We can talk about the detail when it is time to do this. The time range we want to keep affects the detail of the design for this
